cranelift/codegen/src/isa/x86/unwind/winx64.rs

```diff
@@ -183,41 +136,41 @@ pub(crate) fn create_unwind_info(
 
     assert!(found_end);
 
-    if saved_fpr {
-        if static_frame_allocation_size > 240 && saved_fpr {
-            warn!("stack frame is too large ({} bytes) to use with Windows x64 SEH when preserving FPRs. \
-                This is a Cranelift implementation limit, see \
-                https://github.com/bytecodealliance/wasmtime/issues/1475",
-                static_frame_allocation_size);
-            return Err(CodegenError::ImplLimitExceeded);
+    let mut frame_register_offset = 0;
+    if xmm_save_count > 0 {
+        // If there are XMM saves, determine the number of 16-byte slots used for all CSRs (including GPRs)
+        // The "frame register offset" will point at the last slot used (i.e. the last saved FPR)
+        // Assumption: each FPR is stored at a lower address than the previous one
+        let mut last_stack_offset = None;
+        let mut fpr_save_count: u8 = 0;
+        let mut gpr_push_count: u8 = 0;
+        for code in unwind_codes.iter_mut() {
+            match code {
+                UnwindCode::SaveXmm { stack_offset, .. } => {
+                    if let Some(last) = last_stack_offset {
+                        assert!(last > *stack_offset);
+                    }
+                    last_stack_offset = Some(*stack_offset);
+                    fpr_save_count += 1;
+                    *stack_offset = (xmm_save_count - fpr_save_count) as u32 * 16;
+                }
+                UnwindCode::PushRegister { .. } => {
+                    gpr_push_count += 1;
+                }
+                _ => {}
+            }
         }
-        // Only test static frame size is 16-byte aligned when an FPR is saved to avoid
-        // panicking when alignment is elided because no FPRs are saved and no child calls are
-        // made.
-        assert!(
-            static_frame_allocation_size % 16 == 0,
-            "static frame allocation must be a multiple of 16"
-        );
-    }
+        assert_eq!(fpr_save_count, xmm_save_count);
 
-    // Hack to avoid panicking unnecessarily. Because Cranelift generates prologues with RBP at
-    // one end of the call frame, and RSP at the other, required offsets are arbitrarily large.
-    // Windows x64 SEH only allows this offset be up to 240 bytes, however, meaning large
-    // frames are inexpressible, and we cannot actually compile the function. In case there are
-    // no preserved FPRs, we can lie without error and claim the offset to RBP is 0 - nothing
-    // will actually check it. This, then, avoids panics when compiling functions with large
-    // call frames.
-    let reported_frame_offset = if saved_fpr {
-        (static_frame_allocation_size / 16) as u8
-    } else {
-        0
-    };
+        // Account for alignment space when there's an odd number of GPR pushes
+        frame_register_offset = fpr_save_count + ((gpr_push_count + 1) / 2);
```

Comment:
Being `winx64.rs`, I think the assumption that a GPR is half the size of an FPR is workable, but do you think that's worth noting in this comment?

Reply:
I can add a comment.